src/v/wasm/wasmtime.cc

```diff
@@ -540,20 +555,158 @@ class wasmtime_engine : public engine {
+    wasmtime_engine_v2(const wasmtime_engine_v2&) = delete;
+    wasmtime_engine_v2& operator=(const wasmtime_engine_v2&) = delete;
+    wasmtime_engine_v2(wasmtime_engine_v2&&) = delete;
+    wasmtime_engine_v2& operator=(wasmtime_engine_v2&&) = delete;
+    ~wasmtime_engine_v2() override = default;
+
+    ss::future<> start() final {
+        _transform_module_v2.start();
+        co_await create_instance();
+        _main_task = initialize_wasi();
+        co_await _transform_module_v2.await_ready();
+    }
+
+    ss::future<> stop() final {
+        _transform_module_v2.stop(std::make_exception_ptr(
+          wasm_exception("vm was shutdown", errc::engine_shutdown)));
+        co_await std::exchange(_main_task, ss::now());
+        // Deleting the store invalidates the instance and actually frees the
+        // memory for the underlying instance.
+        _store = nullptr;
+        vassert(
+          !_pending_host_function,
+          "pending host functions should be awaited upon before stopping the "
+          "engine");
+        co_return;
+    }
+
+    ss::future<model::record_batch>
+    transform(model::record_batch batch, transform_probe* probe) override {
+        vlog(wasm_log.trace, "Transforming batch: {}", batch.header());
+        if (batch.record_count() == 0) {
+            co_return std::move(batch);
+        }
+        if (batch.compressed()) {
+            batch = co_await storage::internal::decompress_batch(
+              std::move(batch));
+        }
+        ss::future<model::record_batch> fut = co_await ss::coroutine::as_future(
+          invoke_transform(std::move(batch), probe));
+        if (fut.failed()) {
+            probe->transform_error();
+            std::rethrow_exception(fut.get_exception());
+        }
+        co_return std::move(fut).get();
+    }
+
+private:
+    ss::future<> initialize_wasi() {
+        auto* ctx = wasmtime_store_context(_store.get());
+        wasmtime_extern_t start;
+        bool ok = wasmtime_instance_export_get(
+          ctx,
+          &_instance,
+          wasi::preview_1_start_function_name.data(),
+          wasi::preview_1_start_function_name.size(),
+          &start);
+        if (!ok || start.kind != WASMTIME_EXTERN_FUNC) {
+            throw wasm_exception(
+              "Missing wasi _start function", errc::user_code_failure);
+        }
+        vlog(wasm_log.info, "starting wasm vm {}", _meta.name());
+        std::exception_ptr ex;
+        try {
+            co_await call_host_func(ctx, &start.of.func, {}, {});
+        } catch (...) {
+            ex = std::current_exception();
+            vlog(wasm_log.warn, "wasm vm failed: {}", ex);
+        }
+        vlog(wasm_log.info, "wasm vm {} finished", _meta.name());
+        if (!ex) {
+            ex = std::make_exception_ptr(
+              wasm_exception("vm has exited", errc::engine_not_running));
+        }
+        _transform_module_v2.stop(ex);
```

Comment:
does it matter that module::stop() might not do anything with the exception pointer (e.g. the host/guest condition variables there aren't set)? i can't remember the rules like if an exception is thrown but some how never caught? well, i guess this isn't really the same as throwing it, its just an exception_ptr which presumably is benign.

Reply:
> i guess this isn't really the same as throwing it, its just an exception_ptr which presumably is benign.

Yeah my experiments show that it's fine to create an exception_ptr then drop it on the floor.